### ISSUE TITLE
[FEATURE] Execute previous commands

### ISSUE THREAD

**Issue by danthe1st** [gsayson/deacon#7]
**Is your feature request related to a problem? Please describe.**
It is very useful to access previous commands in a shell without going back to the whole line.

**Describe the solution you'd like**
In many shells, there is a built-in `!!` which represents the last executing command. For example, when running `sudo !!` after `apt update`, it runs `sudo apt update`. Just executing `!!` runs the last command again.
Similarly, those shells allow writing `!` followed by text and it then executes the last command that starts with that text (e.g. `sudo !apt` runs `sudo apt update`)

**Describe alternatives you've considered**
It would also be possible to use a different character but that would be surprising to people who used other shells before.

**Additional context**
This should work in scripts as well as in interactive mode (like any syntax feature).
It should also work inside variable substitutions, function calls etc.

**Comment by gsayson** [comment]
Considering that `!` already represents EMBES (exclamation-mark builtin escape syntax; see `docs/error-index.md` error 01), should we just provide some form of special handling for the `!!` case?

I doubt that there would be an executable on the path named `!`.

Also, how would we handle it if there was no previous command executed before (`!!` is the first command in the history)?

**Comment by SaHHiiLL** [comment]
> should we just provide some form of special handling for the `!!` case?

Yes.

> I doubt that there would be an executable on the path named !.

Well, you doubt that but it is entirely possible to make an executable with the name `!`.

> Also, how would we handle it if there was no previous command executed before (!! is the first command in the history)?

just error out in that case, I don't see a reason why it should be any different.  
here's an example of when I cleared my `.zsh_histroy` and ran `!!`

![image](https://user-images.githubusercontent.com/87093947/209924636-1187572e-8c82-4651-9d5a-d11ec1465743.png)

**Comment by gsayson** [comment]
> ... it is entirely possible to make an executable with the name `!`.

Yeah, but there's an insanely low chance that a user will install an executable named `!` and put it on `?PATH?`.
Anyways, I'll just assign error code `02` to `!!` having no command to run.

I'll aim to close this by 0.3.0.

**Comment by danthe1st** [comment]
It should be possible to still execute executables starting with `!`. One possibility would be to quote it (e.g. `'!a'` or `'/bin/!a'`).
